src/sentry/static/sentry/app/views/settings/account/accountSecurity/accountSecurityEnroll.jsx

```diff
@@ -0,0 +1,316 @@
+
+    let data = {
+      phone: this._form.phone,
+      // Send null OTP if we are submitting OTP verification
+      // Otherwise API will think that we are on verification step (e.g. after submitting phone)
+      otp: hasSentCode ? this._form.otp || '' : undefined,
+      // ...((dataModel && dataModel.toJSON()) || {}),
+      secret: authenticator.secret,
+    };
+
+    // Only show loading when submitting OTP
+    this.setState({
+      loading: hasSentCode,
+    });
+
+    if (!hasSentCode) {
+      addMessage(`Sending code to ${data.phone}...`);
+    }
+
+    this.api
+      .requestPromise(`${ENDPOINT}${this.props.params.authId}/enroll/`, {
+        data,
+      })
+      .then(
+        () => {
+          if (!hasSentCode) {
+            // Just successfully finished sending OTP to user
+            this.setState({
+              hasSentCode: true,
+              loading: false,
+              // authenticator: data,
+            });
+            addMessage(`Sent code to ${data.phone}`);
+          } else {
+            // OTP was accepted and SMS was added as a 2fa method
+            this.props.router.push('/settings/account/security/');
+            addSuccessMessage(`Added authenticator ${authenticator.name}`);
+          }
+        },
+        error => {
+          let isSmsInterface = authenticator.id === 'sms';
+
+          // Re-mount because we want to fetch a fresh secret
+          this.remountComponent();
+
+          let errorMessage = this.state.hasSentCode
+            ? 'Incorrect OTP'
+            : 'Error sending SMS';
+          addErrorMessage(errorMessage);
+
+          this.setState({
+            hasSentCode: !isSmsInterface,
+          });
+        }
+      );
+  };
+
+  // Handle u2f device tap
+  handleU2fTap = data => {
+    return this.api
+      .requestPromise(`${ENDPOINT}${this.props.params.authId}/enroll/`, {
+        data: {
+          ...data,
+          ...this._form,
+        },
+      })
+      .then(this.handleEnrollSuccess, this.handleEnrollError);
+  };
+
+  // Currently only TOTP uses this
+  handleSubmit = dataModel => {
+    let {authenticator} = this.state;
+
+    let data = {
+      ...this._form,
+      ...((dataModel && dataModel.toJSON()) || {}),
+      secret: authenticator.secret,
+    };
+
+    this.setState({
+      loading: true,
+    });
+    this.api
+      .requestPromise(`${ENDPOINT}${this.props.params.authId}/enroll/`, {
+        method: 'POST',
+        data,
+      })
+      .then(this.handleEnrollSuccess, this.handleEnrollError);
+  };
+
+  // Handler when we successfully add a 2fa device
+  handleEnrollSuccess = () => {
+    let authenticatorName =
+      (this.state.authenticator && this.state.authenticator.name) || 'Authenticator';
+    this.props.router.push('/settings/account/security');
+    addSuccessMessage(`${authenticatorName} has been added`);
+  };
+
+  // Handler when we failed to add a 2fa device
+  handleEnrollError = () => {
```

Comment:
I'd definitely cancel loading here too, it's easy to type the wrong code and get here